feat: support eslint 9

Smrtnyk opened the pull request:
Also increased major version
 
 I tested this in my repo https://github.com/Smrtnyk/Firetable where I am using eslint 9 and flat config, and it works fine with these changes
 
 I ran the tests:all command it also passes all the tests.

Smrtnyk committed:
feat: support eslint 9

 Also increased major version

getify commented on lib/index.js:
```diff
@@ -110,7 +108,10 @@ module.exports = {
 						// handle "unused" mode
 						if (!noneUnusedMode) {
-							let scope = context.getScope();
+							let sourceCode = context.sourceCode ?? context.getSourceCode();
```
seems like we're checking these things repeatedly (basically, determining ESLint 9+ versus 8 and below)... wondering if we could test once, and centralize...

for example:

```js
var getScope = (context,sourceCode) => {
   getScope = (
      sourceCode.getScope ?
         (context,sourceCode) => sourceCode.getScope() :
         (context,sourceCode) => context.getScope()
   );
   return getScope(context,sourceCode);
```

Here, we lazily do the test the first time `getScope(..)` is called, then overwrite the `getScope(..)` function so it doesn't need to do the checks any more throughout that run.

And we could do similar for the "sourceCode" and "ancestors" tests/evaluations.

Perhaps an approach like this reduces the code AND is *slightly* more performant?

Smrtnyk replied:
I have pushed the changes
Let me know if I misunderstood what you meant

getify commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+.idea
```
not needed for general release

Smrtnyk replied:
will revert

Smrtnyk commented on lib/index.js:
```diff
@@ -3,7 +3,6 @@
 module.exports = {
 	configs: {
 		"getify-says": {
-			plugins: [ "@getify/proper-arrows", ],
```
it was breaking my eslint server in webstorm, because in flat confing plugins is not an array anymore but an object
I had to define plugins now with
```typescript
import arrows from "@getify/eslint-plugin-proper-arrows";
{
      plugins: {
          "@getify/proper-arrows": arrows,
      }
}
```

Smrtnyk commented on lib/index.js:
```diff
@@ -60,7 +59,6 @@ module.exports = {
 			create(context) {
 				var defaultsOnly = context.options.length == 0;
 				var extraOptions = (!defaultsOnly) ? context.options[0] : null;
-				var allUnusedMode = defaultsOnly || !("unused" in extraOptions) || extraOptions.unused === "all";
```
it is a dead code, unused variable
but I will revert it since it is not related to the thing we are fixing

getify commented on lib/index.js:
```diff
@@ -801,3 +810,26 @@ function currentlyInGlobalScope(parserOptions,scope) {
+var getSourceCode = (context) => {
+	getSourceCode = (context) => context.sourceCode ?? context.getSourceCode();
+	return getSourceCode(context);
+};
```
This should instead be:

```js
var getSourceCode = context => {
   getSourceCode = (
      context.sourceCode ?
         context => context.sourceCode : 
         context => context.getSourceCode()
   );
   return getSourceCode(context);
};
```

getify reviewed (changes requested):
The rest LGTM, thanks! :)

getify reviewed (approved):
Code LGTM.

We may need some docs changes... and we also should figure out how v9+ handle the config presets (given that other breakage you were discussing).

Smrtnyk commented:
> Code LGTM.
> 
> We may need some docs changes... and we also should figure out how v9+ handle the config presets (given that other breakage you were discussing).

I can help you with that.
Do you want it to be part of this PR or separate?
I believe this PR is enough to get the people going and unblock them from eslint 9 upgrade (such as myself at the company I work for), and then I can contribute another PR with flat config preset.
I believe that flat config preset could require some restructuring of the plugin object, so it might involve more code changes.
Either way I want to help, so let me know what you prefer :)

getify commented:
removing the preset would be a "breaking change", which sort of defeats the intent here of "fixing" the plugin to keep working without such forking/breakage.

so I would like to at least understand better what other changes are necessary to preserve presets before wrapping this up.

it might well belong in a separate PR, but I think we need that to land before a new "release"... so I just want to have a better sense of the "scope" of such work before we finalize this PR.

also, if this plugin is now used/activated/configured differently in v9+, I want to make sure the docs have a mention of that difference.

Smrtnyk commented:
> removing the preset would be a "breaking change", which sort of defeats the intent here of "fixing" the plugin to keep working without such forking/breakage.
> 
> so I would like to at least understand better what other changes are necessary to preserve presets before wrapping this up.
> 
> it might well belong in a separate PR, but I think we need that to land before a new "release"... so I just want to have a better sense of the "scope" of such work before we finalize this PR.
> 
> also, if this plugin is now used/activated/configured differently in v9+, I want to make sure the docs have a mention of that difference.

I think we can leave your preset as it is
and add a new one maybe called `getify-says/flat`.
I had a bit of a look at how eslint promise plugin did this
https://github.com/eslint-community/eslint-plugin-promise/blob/main/index.js#L51
I tried something like this but there were some exceptions in the eslint server and cli linter, which made me starting to restructure the current plugin object.
I think exceptions came due to some self referencing objects.

How I used this plugin with eslint 9 when testing was like this:
when I removed the plugins array which you aasked about, I continued using `getify-says` preset, but I added proper arrows plugin manually myself to the plugins object in my flat config.
After I reverted the plugins change in `getify-says` preset, I had to remove it from my flat config and manually declare the plugin + rules.

Smrtnyk committed:
feat: support eslint 9

 Add flat config preset

Smrtnyk commented:
I have pushed a separate second commit, that adds a flat config preset
you can see how it is used in here https://github.com/Smrtnyk/Firetable/commit/43ffd66d59c863666622bc76907a28d1fac38b60#diff-a32a0887ed9d1d707bbb3b845b7df7fd40e673c47e7b60a3ebd896b68d3b8839R25

If that is fine for you I can then add some docs for this

Smrtnyk commented:
btw I noticed that there is no CI build on PR branches.
I see that there is a travis config file but I am not able to see the builds.
I would feel more comfortable if there was a build and test github action running when I push to the branch.
I am building and running tests locally, but still.
If you are open for a small PR to setup a github action to run build and test I can contribute

getify commented:
yes, happy to look at any and all contributions, especially helpful ideas like the CI stuff. :)

Smrtnyk commented:
> yes, happy to look at any and all contributions, especially helpful ideas like the CI stuff. :)

added a simple workflow in https://github.com/getify/eslint-plugin-proper-arrows/pull/39
tested it locally using act and docker, seems like it works fine

Smrtnyk commented:
@getify In second commit I have updated the readme and added a flat config entry for the `getify-says` preset
I have tested it in my repo https://github.com/Smrtnyk/Firetable/blob/flat-config-arrow-functions-eslint-plugin/eslint.config.js#L25 and it works nice

getify commented on README.md:
```diff
@@ -80,6 +80,39 @@ The available configuration presets to choose from:
+import arrows from "@getify/eslint-plugin-proper-arrows";
+
+export default [
+    arrows.flatConfigs["CONFIG-PRESET-NAME"],
```
can we add a line-comment at the end of this line, like:

```js
   ...    // i.e., "getify-says", etc
```

getify commented on README.md:
````diff
@@ -80,6 +80,39 @@ The available configuration presets to choose from:
+
+```js
+// eslint.config.js
+import arrows from "@getify/eslint-plugin-proper-arrows";
````
maybe it would be a little more proper to name this import name `properArrows` since that's the name of the plugin? Same with the other instances here in your additions.

getify commented on lib/index.js:
```diff
@@ -1,17 +1,19 @@
 "use strict";
 
-module.exports = {
+var recommendedRulesConfig = {
```
since this variable is only the "getify-says" recommended rules config, let's name it like this:

```js
var presetRulesConfig = {
   "getify-says": { ... }
};
```

And then in the other places we use `recommendedRulesConfig`, let's change those to `presetRulesConfig["getify-says"]`.

This will "future-proof" to if we add other presets besides my own personal opinion.

getify reviewed (changes requested):
This is looking good, appreciate it. Just a few minor tweaks.

Smrtnyk commented:
> This is looking good, appreciate it. Just a few minor tweaks.

tnx for the feedback Kyle
I have force pushed the requested changes to the last commit

getify reviewed (approved):
LGTM!

getify commented:
**NOTE:** I'll be delaying publishing this updated version on npm for a bit, just to give us time to get some feedback from folks if they try it.

chelentos commented:
Hello!

Could you tell please when will be published eslint 9 version of package?
